File: momentra/momentra/Shell/BusinessActive/BusinessTabLoad.swift
import Foundation

enum BusinessTabLoad {
    private static let lock = NSLock()
    private static var inflight: [String: Task<BusinessTabDataCache.PulseTab, Error>] = [:]

    /// Personal-parity pulse load: one bundled /pulse GET (finance + activity preview).
    /// Team Ops capacity/workload are a cheap follow-up and do not block first paint.
    static func loadPulseTab(
        momentId: String,
        fetchTeamOpsMetrics: Bool = false
    ) async throws -> BusinessTabDataCache.PulseTab {
        let owned: (task: Task<BusinessTabDataCache.PulseTab, Error>, owner: Bool) = {
            lock.lock()
            defer { lock.unlock() }
            if let existing = inflight[momentId] { return (existing, false) }
            let task = Task { try await fetchPulseTab(momentId: momentId) }
            inflight[momentId] = task
            return (task, true)
        }()
        defer {
            if owned.owner {
                lock.lock()
                inflight.removeValue(forKey: momentId)
                lock.unlock()
            }
        }
        var tab = try await owned.task.value
        if fetchTeamOpsMetrics {
            tab = await enrichTeamOps(momentId: momentId, tab: tab)
        }
        return tab
    }

    private static func fetchPulseTab(momentId: String) async throws -> BusinessTabDataCache.PulseTab {
        let mark = ShellPerf.start("pulse_tab_ready")
        let pulse = try await APIClient.shared.getBusinessPulse(momentId: momentId)
        let finance = pulse.payload?.finance
        let activities: [APIClient.ActivityItemPayload]
        if let bundled = pulse.payload?.activity {
            activities = bundled
        } else {
            activities = try await APIClient.shared.listBusinessActivity(
                momentId: momentId,
                limit: BusinessTabPrefetch.activityLimit
            )
        }
        let previous = BusinessTabDataCache.peekPulse(momentId)
        let tab = BusinessTabDataCache.PulseTab(
            pulse: pulse,
            finance: finance ?? previous?.finance,
            life: previous?.life,
            activities: activities,
            businessFamily: pulse.businessFamily,
            facetStatus: pulse.status,
            capacity: previous?.capacity,
            workload: previous?.workload
        )
        BusinessTabDataCache.putPulse(momentId, tab)
        ShellPerf.end(mark, extras: ["context": "BUSINESS", "bundled": true])
        return tab
    }

    private static func enrichTeamOps(
        momentId: String,
        tab: BusinessTabDataCache.PulseTab
    ) async -> BusinessTabDataCache.PulseTab {
        if tab.capacity != nil && tab.workload != nil { return tab }
        async let capTask = APIClient.shared.getBusinessCapacity(momentId: momentId)
        async let workTask = APIClient.shared.getBusinessWorkload(momentId: momentId)
        let enriched = BusinessTabDataCache.PulseTab(
            pulse: tab.pulse,
            finance: tab.finance,
            life: tab.life,
            activities: tab.activities,
            businessFamily: tab.businessFamily,
            facetStatus: tab.facetStatus,
            capacity: (try? await capTask) ?? tab.capacity,
            workload: (try? await workTask) ?? tab.workload
        )
        BusinessTabDataCache.putPulse(momentId, enriched)
        return enriched
    }

    static func loadMemoryTab(momentId: String) async throws -> BusinessTabDataCache.MemoryTab {
        let cached = BusinessTabDataCache.peekPulse(momentId)
        async let memoryTask = APIClient.shared.getBusinessMemory(momentId: momentId)
        let memory = try await memoryTask
        let pulse: APIClient.BusinessPulsePayload?
        let finance: APIClient.BusinessFinancePayload?
        if let cachedPulse = cached?.pulse {
            pulse = cachedPulse
            finance = cached?.finance ?? cachedPulse.payload?.finance
        } else {
            let loaded = try await APIClient.shared.getBusinessPulse(momentId: momentId)
            pulse = loaded
            finance = loaded.payload?.finance
        }
        let tab = BusinessTabDataCache.MemoryTab(
            memory: memory,
            pulse: pulse,
            finance: finance,
            life: cached?.life
        )
        BusinessTabDataCache.putMemory(momentId, tab)
        return tab
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/business/shared/BusinessTabLoad.kt
package com.example.momentra.ui.shell.business.shared

import com.example.momentra.data.repository.BusinessSliceRepository
import com.example.momentra.ui.shell.perf.ShellPerf
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import java.util.concurrent.ConcurrentHashMap

const val BUSINESS_PULSE_ACTIVITY_LIMIT = 5

private val pulseInflight =
    ConcurrentHashMap<String, CompletableDeferred<Result<BusinessTabDataCache.PulseTab>>>()

/**
 * Personal-parity pulse load: one bundled /pulse GET (finance + activity preview).
 * Callers should paint [BusinessTabDataCache.peekPulse] before awaiting.
 */
suspend fun loadBusinessPulseTab(
    repository: BusinessSliceRepository,
    momentId: String,
    activityLimit: Int = BUSINESS_PULSE_ACTIVITY_LIMIT,
    fetchTeamOpsMetrics: Boolean = false,
): Result<BusinessTabDataCache.PulseTab> {
    val created = CompletableDeferred<Result<BusinessTabDataCache.PulseTab>>()
    val existing = pulseInflight.putIfAbsent(momentId, created)
    val shared = existing ?: created
    if (existing == null) {
        try {
            created.complete(fetchBundledPulse(repository, momentId, activityLimit))
        } catch (t: Throwable) {
            created.complete(Result.failure(t))
        } finally {
            pulseInflight.remove(momentId, created)
        }
    }
    val base = shared.await()
    if (!fetchTeamOpsMetrics) return base
    return base.mapCatching { tab -> enrichTeamOps(repository, momentId, tab) }
}

private suspend fun fetchBundledPulse(
    repository: BusinessSliceRepository,
    momentId: String,
    activityLimit: Int,
): Result<BusinessTabDataCache.PulseTab> = runCatching {
    val mark = ShellPerf.start("pulse_tab_ready")
    val pulseFacet = repository.getPulse(momentId).getOrThrow()
    val pulse = pulseFacet.payload
    val finance = pulse?.finance
    val activities = pulse?.activity
        ?: repository.getActivity(momentId, limit = activityLimit).getOrThrow().items
    val previous = BusinessTabDataCache.peekPulse(momentId)
    val data = BusinessTabDataCache.PulseTab(
        pulse = pulse,
        finance = finance ?: previous?.finance,
        life = previous?.life,
        activities = activities,
        businessFamily = pulseFacet.businessFamily,
        facetStatus = pulseFacet.status,
        capacity = previous?.capacity,
        workload = previous?.workload,
    )
    BusinessTabDataCache.putPulse(momentId, data)
    ShellPerf.end(mark, mapOf("context" to "BUSINESS", "bundled" to true, "cached" to false))
    data
}

private suspend fun enrichTeamOps(
    repository: BusinessSliceRepository,
    momentId: String,
    tab: BusinessTabDataCache.PulseTab,
): BusinessTabDataCache.PulseTab {
    if (tab.capacity != null && tab.workload != null) return tab
    return coroutineScope {
        val capacityDeferred = async { repository.getCapacity(momentId) }
        val workloadDeferred = async { repository.getWorkload(momentId) }
        val enriched = tab.copy(
            capacity = capacityDeferred.await().getOrNull() ?: tab.capacity,
            workload = workloadDeferred.await().getOrNull() ?: tab.workload,
        )
        BusinessTabDataCache.putPulse(momentId, enriched)
        enriched
    }
}

/**
 * Memory tab load — reuses warm pulse/finance cache when present; only /memory is required.
 */
suspend fun loadBusinessMemoryTab(
    repository: BusinessSliceRepository,
    momentId: String,
): Result<BusinessTabDataCache.MemoryTab> = runCatching {
    coroutineScope {
        val cached = BusinessTabDataCache.peekPulse(momentId)
        val memoryDeferred = async { repository.getMemory(momentId) }
        val pulseDeferred = if (cached?.pulse != null) null else async { repository.getPulse(momentId) }
        val memory = memoryDeferred.await().getOrThrow().payload
        val pulseFacet = pulseDeferred?.await()?.getOrThrow()
        val pulse = cached?.pulse ?: pulseFacet?.payload
        val finance = cached?.finance ?: pulse?.finance ?: pulseFacet?.payload?.finance
        val data = BusinessTabDataCache.MemoryTab(
            memory = memory,
            pulse = pulse,
            finance = finance,
            life = cached?.life,
        )
        BusinessTabDataCache.putMemory(momentId, data)
        data
    }
}

/** Prefetch bundled pulse for Business tab SWR. */
suspend fun prefetchBusinessTabs(repository: BusinessSliceRepository, momentId: String) {
    if (momentId.isBlank()) return
    loadBusinessPulseTab(repository, momentId)
}
